src/Common/tests/Tests/System/Collections/Generic/ArrayBuilderTests.cs

```diff
@@ -0,0 +1,365 @@
+                {
+                    int offset = (builder.Count - 1) - left; // How much we've skipped into the enumerable
+                    Assert.Equal(seed.ElementAt(offset + i), builder[i]);
+                }
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(AddRangeData))]
+        public void AddRange(IEnumerable<T> seed, IEnumerable<T> items)
+        {
+            var builder = CreateBuilderFromSequence(seed);
+
+            builder.AddRange(items.ToArray());
+
+            var expected = seed.Concat(items);
+
+            VerifyBuilderContentsWithForeach(expected, builder);
+            Assert.Equal(expected, builder.ToArray());
+        }
+
+        [Theory]
+        [MemberData(nameof(EnumerableData))]
+        public void GetEnumerator(IEnumerable<T> seed)
+        {
+            var builder = CreateBuilderFromSequence(seed);
+            
+            VerifyBuilderContentsWithForeach(seed, builder);
+        }
+
+        [Theory]
+        [MemberData(nameof(EnumerableData))]
+        public void ToArray(IEnumerable<T> seed)
+        {
+            var builder = CreateBuilderFromSequence(seed);
+
+            int count = builder.Count; // Count needs to be called beforehand.
+            var array = builder.ToArray(); // ToArray should only be called once.
+
+            Assert.Equal(count, array.Length);
+            Assert.Equal(seed, array);
+        }
+
+        [Theory]
+        [MemberData(nameof(CapacityData))]
+        public void UncheckedAdd(int capacity)
+        {
+            Debug.Assert(capacity >= 0);
+
+            var builder = new ArrayBuilder<T>(capacity);
+
+            for (int i = 0; i < capacity; i++)
+            {
+                builder.UncheckedAdd(default(T));
+            }
+            
+            // Count == Capacity now, so attempting to add more should raise
+            // an assert or generate an IndexOutOfRangeException
+            Assert.ThrowsAny<Exception>(() => builder.UncheckedAdd(default(T)));
+        }
+
+        [Theory]
+        [MemberData(nameof(ZeroExtendData))]
+        public void ZeroExtend(IEnumerable<T> seed, int count)
+        {
+            Debug.Assert(count >= 0);
+
+            var builder = CreateBuilderFromSequence(seed);
+
+            builder.ZeroExtend(count);
+
+            var expected = seed.Concat(Enumerable.Repeat(default(T), count));
+
+            VerifyBuilderContentsWithForeach(expected, builder);
+            Assert.Equal(expected, builder.ToArray());
+        }
+
+        public static TheoryData<IEnumerable<T>, IEnumerable<T>> AddRangeData()
+        {
+            var data = new TheoryData<IEnumerable<T>, IEnumerable<T>>();
+
+            var enumerables = EnumerableData().Select(array => array[0]).Cast<IEnumerable<T>>();
+
+            // Pair up each of the enumerables with each of the other enumerables, including itself.
+            // This has O(N^2) complexity depending on how many enumerables there are.
```

Comment:
Have to say I'm getting a bit addicted to generating test data-- `InlineData` seems almost primitive now. :)